src/systemd-sonic-generator/systemd-sonic-generator.cpp

```diff
@@ -385,6 +388,56 @@ static void replace_multi_inst_dep(const char *src) {
     rename(tmp_file_path, src);
 }
 
+static void update_environment(const std::string &src)
+{
+    std::ifstream src_file(src);
+    std::ofstream tmp_file(src + ".tmp");
+    bool has_service_section = false;
+    std::string line;
+
+    // locate the [Service] section
+    while (std::getline(src_file, line)) {
+        tmp_file << line << std::endl;
+        if (line.find("[Service]") != std::string::npos) {
+            has_service_section = true;
+            break;
+        }
+    }
+
+    if (!has_service_section) {
+        tmp_file << "[Service]" << std::endl;
+    }
+
+    std::unordered_map<std::string, std::string> env_vars;
+    env_vars["IS_DPU_DEVICE"] = (smart_switch_dpu ? "true" : "false");
+    env_vars["NUM_DPU"] = std::to_string(num_dpus);
```

Comment:
This is a good approach to solve this issue but i see one setback. If we need to add new env variables, ssg code should be updated. which IMO is not very flexible. I have an idea for generic solution, let me know what you think

1) Add a oneshot service very early in the boot. Read static env variables (Eg: $PLATFORM, $NUM_ASIC, $NUM_DPU, $SONIC_BOOT_TYPE etc) and write them to a common file `/etc/sonic/static-env-variables` 
2) We can leverage ssg to write `EnvironmentFile=/etc/sonic/static-env-variables `option to all the services. making the ssg code minimal and flexible. 
3) We can potentially clean a lot of code under docker_image_ctl with this approach

Reply:
Good idea, I can have a try.

Reply:
I tried it, but I feel it wasn't easy yet.

1. The shell of oneshot service you mentioned might look like the following. But at this time, the database service hasn't ready, so I have to parse the `/host/machine.conf` as same as what systemd-sonic-generator did if I would like to get the variable. The SSG has done this by an efficient function, C function, why should I do it again by a shell?

```shell

SYSTEMD_ENV_FILE="/etc/sonic/static_env"

# Load platform from sonic-cfggen

PLATFORM=${PLATFORM:-`sonic-cfggen -H -v DEVICE_METADATA.localhost.platform`}

echo "PLATFORM='${PLATFORM}'" >> "$SYSTEMD_ENV_FILE"

# Parse environment from platform.json

PLATFORM_JSON=/usr/share/sonic/device/$PLATFORM/platform.json

if [ -f "$PLATFORM_JSON" ]; then

# Environment variables for Smart Switch

    NUM_DPU=$(jq -r '.DPUS | length' $PLATFORM_JSON 2>/dev/null)
    if [[ -z "$NUM_DPU" ]]; then
        NUM_DPU=0
    fi

    jq -e '.DPU' $PLATFORM_JSON >/dev/null
    if [[ $? -eq 0 ]]; then
        IS_DPU_DEVICE="true"
    else
        IS_DPU_DEVICE="false"
    fi

    echo "NUM_DPU='${NUM_DPU}'" >> "$SYSTEMD_ENV_FILE"
    echo "IS_DPU_DEVICE='${IS_DPU_DEVICE}'" >> "$SYSTEMD_ENV_FILE"
fi

```
2. As your proposal, if we want to add new env variables, we have to update the shell script. I don't see any difference or challenge in doing this via SSG. If you feel that C code isn't flexible, I have to say we had an old SSG with python code previously, But we discarded it and rewrote it via C due to the efficiency issues.

Reply:
Yes, script couldn't use DB just like SSG

Hmm, difference is we need to run the script once. Since the oneshot service runs before other services are even started, CPU should be fairly free and should be executed quickly. we do need to benchmark this solution to measure impact. 

Advantage being load on SSG is less, all it does it to add EnvironmentFIle= and with some optimization we don't even need to edit and write to the .service file after the first

Reply:
In any case, i'm okay with the current solution. We might move to the generic oneshot service if required in the future

Reply:
If this PR looks good to you, could you please help to approve it.